features/dd-sdk-android-rum/src/main/kotlin/com/datadog/android/rum/internal/net/RumRequestFactory.kt

```diff
@@ -73,40 +86,90 @@ internal class RumRequestFactory(
 
     private fun buildHeaders(
         requestId: String,
-        clientToken: String,
-        source: String,
-        sdkVersion: String
+        idempotencyKey: String?,
+        context: DatadogContext
     ): Map<String, String> {
-        return mapOf(
-            RequestFactory.HEADER_API_KEY to clientToken,
-            RequestFactory.HEADER_EVP_ORIGIN to source,
-            RequestFactory.HEADER_EVP_ORIGIN_VERSION to sdkVersion,
+        val headers = mutableMapOf(
+            RequestFactory.HEADER_API_KEY to context.clientToken,
+            RequestFactory.HEADER_EVP_ORIGIN to context.source,
+            RequestFactory.HEADER_EVP_ORIGIN_VERSION to context.sdkVersion,
             RequestFactory.HEADER_REQUEST_ID to requestId
         )
+        if (idempotencyKey != null) {
+            headers[RequestFactory.DD_IDEMPOTENCY_KEY] = idempotencyKey
+        }
+        return headers
     }
 
     private fun buildTags(
         serviceName: String,
         version: String,
         sdkVersion: String,
         env: String,
-        variant: String
+        variant: String,
+        executionContext: RequestExecutionContext
     ): String {
         val elements = mutableListOf(
             "${RumAttributes.SERVICE_NAME}:$serviceName",
             "${RumAttributes.APPLICATION_VERSION}:$version",
             "${RumAttributes.SDK_VERSION}:$sdkVersion",
-            "${RumAttributes.ENV}:$env"
+            "${RumAttributes.ENV}:$env",
+            "${RETRY_COUNT_KEY}:${executionContext.attempts}"
         )
 
         if (variant.isNotEmpty()) {
             elements.add("${RumAttributes.VARIANT}:$variant")
         }
+        if (executionContext.previousResponseCode != null) {
+            elements.add("${LAST_FAILURE_STATUS_KEY}:${executionContext.previousResponseCode}")
+        }
 
         return elements.joinToString(",")
     }
 
+    @Suppress("TooGenericExceptionCaught")
+    private fun idempotencyKey(byteArray: ByteArray): String? {
+        try {
+            val digest = MessageDigest.getInstance("SHA-1")
+            val hashBytes = digest.digest(byteArray)
+            return hashBytes.toHexString()
```

Comment:
So each feature has to provide the key on its own? I'm not sure if it is the solution we would like to have. Is it possible to lift it to the core? Or at least have something like `ByteArray.sha1()` declared in the core. How iOS SDK is doing that?

Reply:
I went to the core way but then I've seen that in iOS implementation they did it at the feature level and changed it. I think it makes sense at the feature level as `Logs` and `Traces` will not need this. Otherwise in the core you will have to make sure you do it only for the right features (if/else) and not even sure if we have the feature information in that place to check.

Reply:
They have a function to generate the key in the core https://github.com/DataDog/dd-sdk-ios/blob/dec751b901e17e74bc7f96af56558882cacd9282/DatadogInternal/Sources/Extensions/Data%2BCrypto.swift#L11 and it is applied on the each feature level. We should follow the same approach.

Reply:
The only problem I have with this is that for now we are doing this only for RUM and I am afraid it will stay like that. I have no problem with moving this into core but that's the reason I preferred not to do it. Should we do it now or later in case we really need to share it between features ?

Reply:
I mean for now we can just declare extension method to generate SHA1 in the core and call it from the feature-specific factory. We already have the header declared in the core, isn't it?

Reply:
I get the part with the function to have the extension in the core but don't get the header part ? The header is added here so you only need `ByteArray.generateSha1Fingerprint()` function if we want to go this way. I am still thinking it is overkilled as no other code will ever use this function but will let it here for others to say their opinion. @ambushwork @xgouchet @jonathanmos ?